access/ms-teams/app.go

```diff
@@ -0,0 +1,477 @@
+	_, err := a.pd.Create(ctx, id, func() (PluginData, error) {
+		return PluginData{AccessRequestData: data}, nil
+	})
+
+	// If we succeeded to create PluginData, let's post a messages and save created Teams messages
+	if !trace.IsAlreadyExists(err) {
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		recipients := a.getMessageRecipients(ctx, req)
+
+		if len(recipients) == 0 {
+			log.Warning("No recipients to notify")
+		} else {
+			err = a.postMessages(ctx, recipients, id, data)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	// Update the received reviews
+	reviews := req.GetReviews()
+	if len(reviews) == 0 {
+		return nil
+	}
+
+	err = a.postReviews(ctx, id, reviews)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// onResolvedRequest is called when a request is resolved
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	var tag pd.ResolutionTag
+	state := req.GetState()
+	switch state {
+	case types.RequestState_APPROVED:
+		tag = pd.ResolvedApproved
+	case types.RequestState_DENIED:
+		tag = pd.ResolvedDenied
+	default:
+		logger.Get(ctx).Warningf("Unknown state %v (%s)", state, state.String())
+		return trace.Errorf("Unknown state")
+	}
+	err := a.updateMessages(ctx, req.GetName(), tag, req.GetResolveReason(), req.GetReviews())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// onDeleteRequest gets called when a request is deleted
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.updateMessages(ctx, reqID, pd.ResolvedExpired, "", nil)
+}
+
+// postMessages posts initial Teams messages
+func (a *App) postMessages(ctx context.Context, recipients []string, id string, data pd.AccessRequestData) error {
+	teamsData, err := a.bot.PostMessages(ctx, recipients, id, data)
+	if err != nil {
+		if len(teamsData) == 0 {
+			return trace.Wrap(err)
+		}
+
+		logger.Get(ctx).WithError(err).Error("Failed to post one or more messages to MS Teams")
+	}
+
+	for _, data := range teamsData {
+		logger.Get(ctx).WithFields(logger.Fields{
+			"id":        data.ID,
+			"timestamp": data.Timestamp,
+			"recipient": data.RecipientID,
+		}).Info("Successfully posted to MS Teams")
+	}
+
+	// Let's update sent messages data
+	_, err = a.pd.Update(ctx, id, func(existing PluginData) (PluginData, error) {
+		existing.TeamsData = teamsData
+		return existing, nil
+	})
+
+	return trace.Wrap(err)
+}
+
+// postReviews updates a message with reviews
+func (a *App) postReviews(ctx context.Context, id string, reviews []types.AccessReview) error {
+	pluginData, err := a.pd.Update(ctx, id, func(existing PluginData) (PluginData, error) {
+		teamsData := existing.TeamsData
+		if len(teamsData) == 0 {
+			return PluginData{}, trace.CompareFailed("existing teamsData is empty")
+		}
+
+		count := len(reviews)
+		oldCount := existing.ReviewsCount
+		if oldCount >= count {
+			return PluginData{}, trace.CompareFailed("reviews are sent already")
+		}
```

Comment:
If the reviews are already sent, won't the backoff attempts to CompareAndSwap also fail? Should we just return an error other than CompareFailed to abort the update?
```suggestion
		if oldCount >= count 
			return PluginData{}, trace.AlreadyExists("reviews are sent already")
		}
```

Reply:
@Joerger Thank you for your thoroughness!